File: Java/152. Maximum Product Subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int min, max, res;
        max = min = res = nums[0];
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] < 0) {
                int tmp = min;
                min = max;
                max = tmp;                
            }                        
            max = Math.max(nums[i], nums[i] * max);
            min = Math.min(nums[i], nums[i] * min);            
            if (max > res) res = max;                        
        }
        return res;
    }
}

File: Python/152. Maximum Product Subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if nums is None or len(nums) == 0:
            return None
        ans = nums[0]
        imax = ans
        imin = ans
        for num in nums[1:]:
            if num < 0:
                tmp = imax
                imax = imin
                imin = tmp

            imax = max(num, num * imax)
            imin = min(num, num * imin)

            ans = max(imax, ans)
        return ans
